src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -38,28 +40,52 @@ double NaiveKMeans<MetricType, MatType>::Iterate(const arma::mat& centroids,
   counts.zeros(centroids.n_cols);
 
   // Find the closest centroid to each point and update the new centroids.
-  for (size_t i = 0; i < dataset.n_cols; i++)
+  // Computed in parallel over the complete dataset
+  #pragma omp parallel
   {
-    // Find the closest centroid to this point.
-    double minDistance = std::numeric_limits<double>::infinity();
-    size_t closestCluster = centroids.n_cols; // Invalid value.
+    // The current state of the K-means is private for each thread
+    arma::mat localCentroids(centroids.n_rows, centroids.n_cols,
+                             arma::fill::zeros);
+    arma::Col<size_t> localCounts(centroids.n_cols, arma::fill::zeros);
 
-    for (size_t j = 0; j < centroids.n_cols; j++)
+#ifdef _WIN32
+  // Tiny workaround: Visual Studio only implements OpenMP 2.0, which doesn't
+  // support unsigned loop variables. If we're building for Visual Studio, use
+  // the intmax_t type instead.
+  #pragma omp for
+    for (intmax_t i = 0; i < (intmax_t) dataset.n_cols; ++i)
+#else
+  #pragma omp for
+    for (size_t i = 0; i < dataset.n_cols; ++i)
+#endif
     {
-      const double distance = metric.Evaluate(dataset.col(i), centroids.col(j));
+      // Find the closest centroid to this point.
+      double minDistance = std::numeric_limits<double>::infinity();
+      size_t closestCluster = centroids.n_cols; // Invalid value.
 
-      if (distance < minDistance)
+      for (size_t j = 0; j < centroids.n_cols; j++)
       {
-        minDistance = distance;
-        closestCluster = j;
+        const double distance = metric.Evaluate(dataset.col(i),
+                                                centroids.unsafe_col(j));
+        if (distance < minDistance)
+        {
+          minDistance = distance;
+          closestCluster = j;
+        }
       }
-    }
 
-    Log::Assert(closestCluster != centroids.n_cols);
+      Log::Assert(closestCluster != centroids.n_cols);
 
-    // We now have the minimum distance centroid index.  Update that centroid.
-    newCentroids.col(closestCluster) += arma::vec(dataset.col(i));
-    counts(closestCluster)++;
+      // We now have the minimum distance centroid index.  Update that centroid.
+      localCentroids.unsafe_col(closestCluster) += dataset.col(i);
+      localCounts(closestCluster)++;
+    }
+    // Combine calculated state from each thread
+    #pragma omp critical
```

Comment:
I made a few measurements and found that the time spent in the critical section by a thread is quite varied, but small. With a sample size of 8000 measurements, the average time was around 365ns. This measurements is quite shaky, as the resolution of the clock on my system is around 20ns. The maximum time spent was around 82us. But this is too low to affect the runtime significantly.